extension/persistence/relational-jdbc/src/main/java/org/apache/polaris/extension/persistence/relational/jdbc/JdbcBasePersistenceImpl.java

```diff
@@ -127,88 +175,13 @@ public void writeEntity(
-              // lookup by name
-              EntityNameLookupRecord exists =
-                  lookupEntityIdAndSubTypeByName(
-                      callCtx,
-                      entity.getCatalogId(),
-                      entity.getParentId(),
-                      entity.getTypeCode(),
-                      entity.getName());
-              if (exists != null) {
-                throw new EntityAlreadyExistsException(entity);
-              }
-              String query;
-              if (originalEntities == null || originalEntities.get(i) == null) {
-                try {
-                  query = generateInsertQuery(modelEntity, realmId);
-                  statement.executeUpdate(query);
-                } catch (SQLException e) {
-                  if ((datasourceOperations.isConstraintViolation(e)
-                      || datasourceOperations.isAlreadyExistsException(e))) {
-                    throw new EntityAlreadyExistsException(entity, e);
-                  } else {
-                    throw new RuntimeException(
-                        String.format("Failed to write entity due to %s", e.getMessage()), e);
-                  }
-                }
-              } else {
-                Map<String, Object> params =
-                    Map.of(
-                        "id",
-                        originalEntities.get(i).getId(),
-                        "catalog_id",
-                        originalEntities.get(i).getCatalogId(),
-                        "entity_version",
-                        originalEntities.get(i).getEntityVersion(),
-                        "realm_id",
-                        realmId);
-                query = generateUpdateQuery(modelEntity, params);
-                try {
-                  int rowsUpdated = statement.executeUpdate(query);
-                  if (rowsUpdated == 0) {
-                    throw new RetryOnConcurrencyException(
-                        "Entity '%s' id '%s' concurrently modified; expected version %s",
-                        entity.getName(),
-                        entity.getId(),
-                        originalEntities.get(i).getEntityVersion());
-                  }
-                } catch (SQLException e) {
-                  throw new RuntimeException(
-                      String.format("Failed to write entity due to %s", e.getMessage()), e);
-                }
-              }
-            }
-            return true;
-          });
-    } catch (SQLException e) {
-      throw new RuntimeException(
-          String.format(
-              "Error executing the transaction for writing entities due to %s", e.getMessage()),
-          e);
+  private int execute(Object executor, String query) throws SQLException {
+    if (executor instanceof Statement) {
+      return ((Statement) executor).executeUpdate(query);
+    } else if (executor instanceof DatasourceOperations) {
+      return ((DatasourceOperations) executor).executeUpdate(query);
+    } else {
```

Comment:
Would it not be simper to make `writeEntity` use `runWithinTransaction`?

Reply:
I think we want to re:use connection right, or are you suggeting running a transaction within another transaction ?

Reply:
I think perfer the current way, rather than running transaction within another transaction, though i do agree may be i need to think how can i model the transaction and connection sharing

Reply:
Where do you see nested transactions? It's always one transactions with one or many statements.

Reply:
I see, what you mean now, is just to get rid of this if else ? that should be fine, i preferred not starting a transaction for this atomic operation `writeEntitity`
what i understood was to call writeEntity from writeEntities